README.md
<h1>미니게임 천국</h1>
<img src = "https://user-images.githubusercontent.com/102408620/204138800-a562e164-caac-4b5f-b975-3632f6c1a818.PNG">

<b>제작: yoots50, JYLeem</b><br>

상명대학교 알고리즘과게임콘텐츠 강의의 팀프로젝트 파일입니다.<br>
Python을 이용하여 만들 수 있는 게임들을 모은 미니게임 천국 프로그램입니다.


<h1>프로그램에 존재하는 게임 목록</h1>

<h2>1. 틱택토</h2>
<img src = "https://user-images.githubusercontent.com/102408620/204139233-838c5ea3-95e6-4d29-a219-c6389b1bd4c9.PNG">

<h2>2. 베스킨 라빈스 31</h2>
<img src = "https://user-images.githubusercontent.com/102408620/204139242-b1cbf8f7-6627-4bc2-9cbe-679096a41f6c.PNG">

<h2>3. 가까운 숫자 말하기</h2>
<img src = "https://user-images.githubusercontent.com/102408620/204139255-34c382c4-ae81-4c4e-ba23-0d84b970dec3.PNG">

<h2>4. 사다리타기</h2>
<img src = "https://user-images.githubusercontent.com/102408620/204139264-61155b1a-35bb-43f4-8d85-e6bc5e6d4a90.PNG">

<h2>5. 오목</h2>
<img src = "https://user-images.githubusercontent.com/102408620/204139280-284f3eb6-4cdf-41cd-aa0b-3489993c6976.PNG">

<h2>6. Connect4</h2>
<img src = "https://user-images.githubusercontent.com/102408620/204139350-120d8bf7-4c7c-4dbc-8488-fe3a326c8c2f.PNG">

<h1>!주의사항!</h1>

Windows 운영체제를 기반으로 만들어져 있기에 다른 운영체제에선 오목, connect4등의 게임화면이 깨져서 보일 수 있습니다.


change_log.txt
zip에서 수정한거
1. 데코레이터 이용해서 함수 재사용성 증가시킴
2. os.system('ctl')로 메뉴를 선택할 때 마다 출력된 문장들을 초기화 시킴 => python IDLE안됨, 명령창에서 실행(파일 더블클릭)하면 가능
3. tictac.py랑 tictac_main.py합침
4. 틱택토 2인 플레이에서 예외처리 오류 수정

찐막에서 수정한거
1. os.exit() => sys.exit()

찐찐막에서 수정한거
1. 모든 *_main.py에 메인 화면 선택지 추가
2. 모든 *_main.py의 데코레이터 수정
3. 오목 추가
4. 게임 도중 로고가 보이게 수정
5. printLogo()함수를 만들어 재사용성 증가시킴

찐찐찐막에서 수정한거
1. closeNumber_main.py catch 구문 case _: 추가 및 잡다한거 수정
2. 모든 printLogo() 및 printLayout() 코드가 마지막에 엔터를 치도록 수정
3. connect4_main.py 생성 및 main.py에 추가
4. main.py case구문 print() 및 os.system('cls') 제거

찐찐찐찐막에서 수정한거
1. connect4_main.py 오류 수정

찐찐찐찐찐막에서 수정한거
1. mac os에서 os.system('cls')가 안되던거 수정
2. tictac_1p, tictac_2p에서 좌표가 아닌 a~i를 선택하게 수정
3. ledder()가 입력값을 받아 결과를 출력하는 식으로 수정

2022/11/13 수정사항
1. tictac_main.py에 clear()수정
2. br31 choice(func)수정
3. tictac_1p(), tictac_2p()에서 이미 1또는 2가 있는 곳의 위치를 입력할시 예외처리 수정
4. ladder()에서 -1을 입력하면 바로 choice(func)가 작동하도록 수정

2022/11/21 수정사항
1. main.py를 제외한 *.py 파일에 #아래의 출처에서 코드를 수정함, #출처: https://stackoverflow.com/questions/22885780/python-clear-the-screen 추가
2. 사다리타기에서 순서를 다르게 한 숫자씩 적어 결과를 보면 모든 결과를 보아도 게임이 끝나지 않았던 버그 수정

main.py

import sys
from tictac_main import tictac_main
from ladder_main import ladder_main
from br31_main import br31_main
from closeNumber_main import closeNumber_main
from omok_main import omok_main
from connect4_main import connect4_main

while True:
    print(" __  __ _       _")
    print("|  ⧵/  (_)_ __ (_) __ _  __ _ _ __ ___   ___")
    print("| |⧵/| | | '_ ⧵| |/ _` |/ _` | '_ ` _ ⧵ / _ ⧵")
    print("| |  | | | | | | | (_| | (_| | | | | | |  __/")
    print("|_|  |_|_|_| |_|_|⧵__, |⧵__,_|_| |_| |_|⧵___|")
    print("                  |___/")
    print(" _   _")
    print("| | | | ___  __ ___   _____ _ __")
    print("| |_| |/ _ ⧵/ _` ⧵ ⧵ / / _ ⧵ '_ ⧵")
    print("|  _  |  __/ (_| |⧵ V /  __/ | | |")
    print("|_| |_|⧵___|⧵__,_| ⧵_/ ⧵___|_| |_|")
    print("----------안녕하세요! 미니게임 천국입니다.----------")
    print("플레이 하실 게임을 골라주세요!")
    game_type = int(input("[1. Tic tac toe] [2. 배스킨 라벤스 31] [3. 가까운 숫자 말하기] [4. 사다리타기] [5. 오목] [6. Connect4] [0. 종료]: "))
    game_type_list = [1, 2, 3, 4, 5, 6, 0]

    while True:
        if game_type not in game_type_list:
            print("올바르게 입력하여 주세요. \n")
            game_type = int(input("[1. Tic tac toe] [2. 배스킨 라벤스 31] [3. 가까운 숫자 말하기] [4. 사다리타기] [5. 오목] [6. Connect4] [0. 종료]: "))
        else:
            break
    match game_type:
        case 1:
            tictac_main()
        case 2:
            br31_main()
        case 3:
            closeNumber_main()
        case 4:
            ladder_main()
        case 5:
            omok_main()
        case 6:
            connect4_main()
        case 0:
            sys.exit()





omok_main.py

import os

#아래의 출처에서 코드를 수정함
#출처: https://stackoverflow.com/questions/22885780/python-clear-the-screen
from platform   import system as system_name 
def clear_screen(): 
    command = 'cls' if system_name().lower().startswith('win') else 'clear'
    os.system(command)

def choice(func):
    def wrapper():
        if func():
            clear_screen()
            return 0
        ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        while ch != '1' and ch != '2':
            print('다시 입력하세요.')
            ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        else:
            if ch == '1':
                clear_screen()
                @choice
                def func1():
                    func()
                return func1()
            else:
                clear_screen()
    return wrapper

def printLogo():
    clear_screen()
    print("                       _")
    print("  ___  _ __ ___   ___ | | __")
    print(" / _ ⧵| '_ ` _ ⧵ / _ ⧵| |/ /")
    print("| (_) | | | | | | (_) |   <")
    print(" ⧵___/|_| |_| |_|⧵___/|_|⧵_⧵")
    print("")
    
def printLayout(arr, n, m):
    printLogo()
    print("   ①  ②  ③  ④  ⑤  ⑥  ⑦  ⑧  ⑨  ⑩  ⑪")
    print("")
    for i in range(n):
        if i == 0:
            print("①", end = ' ')
        elif i == 2:
            print("②", end = ' ')
        elif i == 4:
            print("③", end = ' ')
        elif i == 6:
            print("④", end = ' ')
        elif i == 8:
            print("⑤", end = ' ')
        elif i == 10:
            print("⑥", end = ' ')
        elif i == 12:
            print("⑦", end = ' ')
        elif i == 14:
            print("⑧", end = ' ')
        elif i == 16:
            print("⑨", end = ' ')
        elif i == 18:
            print("⑩", end = ' ')
        elif i == 20:
            print("⑪", end = ' ')
        else:
            print("  ", end = ' ')
        for j in range(m):
            print(arr[i][j], end = "")
        print("")
    print("")

def isFull(arr, n, m):
    fcnt = 0
    for i in range(0, n, 2):
        for j in range(0, m, 2):
            if arr[i][j] == '●' or arr[i][j] == '○':
                fcnt += 1
    if fcnt == (m + 1) * (n + 1):
        return 1
    return 0

def isEnd(arr, n, m):
    for i in range(0, n, 2):
        for j in range(0, m, 2):
            if arr[i][j] == '●':
                if isLine(arr, i, j):
                    return 1
            elif arr[i][j] == '○':
                if isLine(arr, i, j):
                    return 2
    return 0

def isLine(arr, i, j):
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i][j + k]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i][j - k]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i + k][j]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i - k][j]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i + k][j + k]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i - k][j - k]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i - k][j + k]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 10, 2):
            if arr[i][j] == arr[i + k][j - k]:
                cnt += 1
        if cnt == 4:
            return 1
    except IndexError:
        pass
    return 0
    
def clear(arr, n, m):
    for i in range(n):
        if i % 2:
            arr[i][0] = '│'
            arr[i][m - 1] = '│'
        else:
            arr[i][0] = '├'
            arr[i][m - 1] = '┤'
        for j in range(m):
            if i % 2 and not j % 2:
                arr[i][j] = '│'
        for j in range(m):
            if not j % 2:
                arr[0][j] = '┬'
                arr[n - 1][j] = '┴'
            else:
                arr[0][j] = '───'
                arr[n - 1][j] = '───'
                if not i % 2:
                    arr[i][j] = '───'
                else:
                    if i < n - 1 and j < m - 1:
                        arr[i + 1][j + 1] = '┼'
    arr[0][0], arr[0][m - 1], arr[n - 1][0], arr[n - 1][m - 1] = '┌', '┐', '└', '┘'

def omok():
    #arr[n][m]
    n = 10 #세로
    m = 10 #가로
    m = 2 * m + 1
    n = 2 * n + 1
    arr = [["   "] * m for i in range(n)]
    clear(arr, n, m)
    while True:
        for i in range(0, 2):
            printLayout(arr, n, m)
            p = input(f"플레이어 {i + 1}, 놓을 곳의 좌표를 i j형식으로 적으세요: ")
            while True:
                try:
                    x = 2 * (int(p.split(" ")[0]) - 1)
                    y = 2 * (int(p.split(" ")[1]) - 1)
                    while (arr[x][y] == '●' or arr[x][y] == '○'):
                        printLayout(arr, n, m)
                        p = input("이미 1 또는 2가 있습니다. 놓을 곳의 좌표를 i j형식으로 적으세요: ")
                        x = 2 * (int(p.split(" ")[0]) - 1)
                        y = 2 * (int(p.split(" ")[1]) - 1)
                    break 
                except BaseException:
                    printLayout(arr, n, m)
                    p = input("오류: 숫자가 아니거나 범위를 벗어남, 다시 적어주세요: ")       
            x = 2 * (int(p.split(" ")[0]) - 1)
            y = 2 * (int(p.split(" ")[1]) - 1)
            if i == 0:
                arr[x][y] = '●'
                if y < 20:
                    arr[x][y + 1] = '──'
                if y > 1:
                    arr[x][y - 1] = '── '
                if arr[x][y - 2] == '●' or arr[x][y - 2] == '○':
                    arr[x][y - 1] = '─ '
            else:
                arr[x][y] = '○'
                if y < 20:
                    arr[x][y + 1] = '──'
                if y > 1:
                    arr[x][y - 1] = '── '
                if arr[x][y - 2] == '●' or arr[x][y - 2] == '○':
                    arr[x][y - 1] = '─ '
            if isFull(arr, n, m) and not isEnd(arr, n, m):
                printLayout(arr, n, m)
                print("무승부")
                return 0
            elif isEnd(arr, n, m):
                printLayout(arr, n, m)
                print(f"게임 끝 승자: {isEnd(arr, n, m)}")
                return 0
            else:
                pass

@choice            
def omok_main():
    printLogo()
    player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    while player_mode != "1" and player_mode != "2":
        printLogo()
        print("잘못 입력하셨습니다. 다시 입력해 주세요.")
        player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    else:
        if player_mode == "1":
            clear_screen()
            omok()
            return 0
        elif player_mode == "2":
            return 1

if __name__ == "__main__":
    #print("격자무늬 만드는 함수입니다.")
    #m = 10#int(input("가로 크기 입력: "))
    #n = 10#int(input("세로 크기 입력: "))
    #omok(m, n)
    omok_main()
    input()
    


ladder_main.py

import random
import os

#아래의 출처에서 코드를 수정함
#출처: https://stackoverflow.com/questions/22885780/python-clear-the-screen
from platform   import system as system_name 
def clear_screen(): 
    command = 'cls' if system_name().lower().startswith('win') else 'clear'
    os.system(command)

def choice(func):
    def wrapper():
        if func():
            clear_screen()
            return 0
        ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        while ch != '1' and ch != '2':
            print('다시 입력하세요.')
            ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        else:
            if ch == '1':
                clear_screen()
                @choice
                def func1():
                    func()
                return func1()
            else:
                clear_screen()
    return wrapper


def printLayout(n, arr):
    for i in range(n):
        print(f"{i + 1}", end = ' ')
    print("")
    for i in range(10):
        for j in range(n * 2):
            print(arr[i][j], end = '')
        print("")
    for i in range(n):
        print(f"{i + 1}", end = ' ')
    print("")

def printLogo():
    clear_screen()
    print(" _           _     _")
    print("| | __ _  __| | __| | ___ _ __")
    print("| |/ _` |/ _` |/ _` |/ _ ⧵ '__|")
    print("| | (_| | (_| | (_| |  __/ |")
    print("|_|⧵__,_|⧵__,_|⧵__,_|⧵___|_|")
    print("")

@choice
def ladder_main():
    printLogo()
    player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    while player_mode != "1" and player_mode != "2":
        printLogo()
        print("잘못 입력하셨습니다. 다시 입력해 주세요.")
        player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    else:
        if player_mode == "1":
            clear_screen()
            ladder()
            return 0
        elif player_mode == "2":
            return 1
            
def ladder():
    printLogo()
    n = input("사람의 수를 적어주세요: ")
    while True:
        try:
            n = int(n)
            break
        except BaseException:
            printLogo()
            n = input("숫자로 입력하세요.")
    printLogo()
    arr = [[0] * n * 2 for i in range(10)] #arr[10][2 * n]
    for i in range(10):
        for j in range(n * 2):
            if j % 2:
                arr[i][j] = " "
            else:
                arr[i][j] = "|"
    for i in range(1, 2 * n - 1, 2):
        for j in range(2):
            y = random.randrange(0, 10)
            while True:
                if i == 1:
                    if arr[y][i + 2] == '-':
                        y = random.randrange(0, 10)
                    else:
                        break
                elif i == 2 * n - 2:
                    if arr[y][i - 2] == '-':
                        y = random.randrange(0, 10)
                    else:
                        break
                else:
                    if arr[y][i + 2] == '-':
                        y = random.randrange(0, 10)
                    elif arr[y][i - 2] == '-':
                        y = random.randrange(0, 10)
                    else:
                        break
            arr[y][i] = "-"
    printLayout(n, arr)
    temp = []
    done = False
    while not done:
        ch = input("모든 결과를 한번에 보시려면 0, 스킵하시려면 -1, 그렇지 않으면 원하는 결과의 숫자를 입력하세요: ")
        while ch not in [str(i - 1) for i in range(int(n + 2))]:
            print("다시 입력하세요.")
            ch = input("모든 결과를 한번에 보시려면 0, 스킵하시려면 -1, 그렇지 않으면 원하는 결과의 숫자를 입력하세요: ")
        if ch == "0":
            for i in range(0, 2 * n, 2):
                x = i
                y = 0
                while True:
                    if x == 0:
                        if arr[y + 1][x + 1] == '-':
                            x += 2
                        y += 1
                    elif x == 2 * n - 2:
                        if arr[y + 1][x - 1] == '-':
                            x -= 2
                        y += 1
                    else:
                        if arr[y + 1][x - 1] == '-':
                            x -= 2
                        elif arr[y + 1][x + 1] == '-':
                            x += 2
                        y += 1
                    if y == 9:
                        print(f"{i // 2 + 1}의 목적지: {x // 2 + 1}")
                        break
            done = True
        elif ch == "-1":
            return 0
        else:
            x = (int(ch) - 1) * 2
            y = 0
            while True:
                if x == 0:
                    if arr[y + 1][x + 1] == '-':
                        x += 2
                    y += 1
                elif x == 2 * n - 2:
                    if arr[y + 1][x - 1] == '-':
                        x -= 2
                    y += 1
                else:
                    if arr[y + 1][x - 1] == '-':
                        x -= 2
                    elif arr[y + 1][x + 1] == '-':
                        x += 2
                    y += 1
                if y == 9:
                    print(f"{int(ch)}의 목적지: {x // 2 + 1}")
                    break
            temp.append(ch)
            temp.sort()
            if temp == [str(i + 1) for i in range(int(n))]:
                done = True
            
if __name__ == "__main__":
    ladder_main()


tictac_main.py

import os
import random

#아래의 출처에서 코드를 수정함
#출처: https://stackoverflow.com/questions/22885780/python-clear-the-screen
from platform   import system as system_name 
def clear_screen(): 
    command = 'cls' if system_name().lower().startswith('win') else 'clear'
    os.system(command)

arr = [["a","b","c"],
       ["d","e","f"],
       ["g","h","i"]]

origin = [["a","b","c"],
          ["d","e","f"],
          ["g","h","i"]]

def printLogo():
    clear_screen()
    print(" _   _      _             _ ")
    print("| |_(_) ___| |_ ____  ___| |_ ____   ___ ")
    print("| __| |/ __| __/ _  |/ __| __/  _ ＼/ _ ＼")
    print("| |_| | (__| || (_| | (__| | | (_) || __/")
    print("＼__|_|＼__|＼_＼___|＼__|＼_＼____/＼___|")
    print("")

def printLayout():
    printLogo()
    print("+---+---+---+")
    for i in range(0, 3):
        print("|", end = '')
        for j in range(0, 3):
            print(f" {arr[i][j]} ", end = '|')
        print("\n+---+---+---+")
    print("")

def isLine():
    global cnt1, cnt2
    if cnt1 == 3:
        return 1
    elif cnt2 == 3:
        return 2
    else:
        cnt1 = 0
        cnt2 = 0

def isEnd():
    global cnt1, cnt2
    cnt1 = 0
    cnt2 = 0
    for i in range(0, 3):
        if arr[i][i] == 1:
            cnt1 += 1
        elif arr[i][i] == 2:
            cnt2 += 1
        else:
            pass
    if isLine():
        return isLine()
    for i in range(0, 3):
        if arr[i][2 - i] == 1:
            cnt1 += 1
        elif arr[i][2 - i] == 2:
            cnt2 += 1
        else:
            pass
    if isLine():
        return isLine()
    for i in range(0, 3):
        for j in range(0, 3):
            if arr[i][j] == 1:
                cnt1 += 1
            elif arr[i][j] == 2:
                cnt2 += 1
            else:
                pass
        if isLine():
            return isLine()
    for i in range(0, 3):
        for j in range(0, 3):
            if arr[j][i] == 1:
                cnt1 += 1
            elif arr[j][i] == 2:
                cnt2 += 1
            else:
                pass
        if isLine():
            return isLine()
    return 0

def isFull():
    fcnt = 0
    for i in range(0, 3):
        for j in range(0, 3):
            if arr[i][j] in [1, 2]:
                fcnt += 1
    if fcnt == 9:
        return 1
    return 0

def computer():
    x = random.randrange(0, 3)
    y = random.randrange(0, 3)
    while (arr[x][y] == 1 or arr[x][y] == 2) :
        x = random.randrange(0, 3)
        y = random.randrange(0, 3)
        if isFull():
            return 0
    arr[x][y] = 2

def clear(arr):
    for i in range(3):
        for j in range(3):
            arr[i][j] = origin[i][j]

def tictac_1p():
    while True:
        done = True
        printLayout()
        p = input("놓을 곳의 좌표를 적으세요: ")
        while done:
            while not p in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i']:
                printLayout()
                p = input("놓을 곳의 좌표를 다시 적으세요: ")
            for j in range(3):
                if p in arr[j]:
                    x = j
                    y = arr[j].index(p)
            if not arr[x][y] in [1, 2]:
                done = False
            else:
                printLayout()
                p = input("이미 칸이 찼습니다. 다시 적으세요: ")
        arr[x][y] = 1
        computer()
        if isFull() and not isEnd():
            printLayout()
            print("무승부")
            clear(arr)
            return 0
        elif isEnd():
            printLayout()
            print(f"게임 끝 승자: {isEnd()}")
            clear(arr)
            return 0
        else:
            pass

def tictac_2p():
    while True:
        for i in range(0, 2):
            done = True
            printLayout()
            p = input("놓을 곳의 좌표를 적으세요: ")
            while done:   
                while not p in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i']:
                    printLayout()
                    p = input("놓을 곳의 좌표를 다시 적으세요: ")
                for j in range(3):
                    if p in arr[j]:
                        x = j
                        y = arr[j].index(p)
                if not arr[x][y] in [1, 2]:
                    done = False
                else:
                    printLayout()
                    p = input("이미 칸이 찼습니다. 다시 적으세요: ")
            arr[x][y] = i + 1
            if isFull() and not isEnd():
                printLayout()
                print("무승부")
                clear(arr)
                return 0
            elif isEnd():
                printLayout()
                print(f"게임 끝 승자: {isEnd()}")
                clear(arr)
                return 0
            else:
                pass

def choice(func):
    def wrapper():
        if func():
            clear_screen()
            return 0
        ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        while ch != '1' and ch != '2':
            print('다시 입력하세요.')
            ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        else:
            if ch == '1':
                clear_screen()
                @choice
                def func1():
                    func()
                return func1()
            else:
                clear_screen()
    return wrapper

@choice
def tictac_main():
    while(True):
        printLogo()
        player_mode = input("[1. 싱글플레이] [2. 멀티 플레이] [3. 메인 화면]\n선택지를 입력: ")
        while player_mode != "1" and player_mode != "2" and player_mode != "3":
            printLogo()
            print("잘못 입력하셨습니다. 다시 입력해 주세요.")
            player_mode = input("[1. 싱글플레이] [2. 멀티 플레이] [3. 메인 화면]\n선택지를 입력: ")
        else:
            if player_mode == "1":
                print("[싱글 플레이]")
                tictac_1p()
                break
            elif player_mode == "2":
                print("[멀티 플레이]")
                tictac_2p()
                break
            elif player_mode == "3":
                return 1
            

if __name__ == "__main__":
    tictac_main()


connect4_main.py
import os

#아래의 출처에서 코드를 수정함
#출처: https://stackoverflow.com/questions/22885780/python-clear-the-screen
from platform   import system as system_name 
def clear_screen(): 
    command = 'cls' if system_name().lower().startswith('win') else 'clear'
    os.system(command)
    
def choice(func):
    def wrapper():
        if func():
            clear_screen()
            return 0
        ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        while ch != '1' and ch != '2':
            print('다시 입력하세요.')
            ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        else:
            if ch == '1':
                clear_screen()
                @choice
                def func1():
                    func()
                return func1()
            else:
                clear_screen()
    return wrapper

def printLogo():
    clear_screen()
    print("                                 _   _  _")
    print("  ___ ___  _ __  _ __   ___  ___| |_| || |")
    print(" / __/ _ ⧵| '_ ⧵| '_ ⧵ / _ ⧵/ __| __| || |_")
    print("| (_| (_) | | | | | | |  __/ (__| |_|__   _|")
    print(" ⧵___⧵___/|_| |_|_| |_|⧵___|⧵___|⧵__|  |_|")
    print("")

def printLayout(arr, n, m):
    printLogo()
    print("  ①  ②  ③  ④  ⑤  ⑥  ⑦")
    print("")
    for i in range(n):
        for j in range(m):
            print(arr[i][j], end = "")
        print("")
    print("")

def isFull(arr, n, m):
    fcnt = 0
    for i in range(1, n, 2):
        for j in range(1, m, 2):
            if arr[i][j] == ' ●' or arr[i][j] == ' ○':
                fcnt += 1
    if fcnt == (m + 1) * (n + 1):
        return 1
    return 0

def isLine(arr, i, j):
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i][j + k]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i][j - k]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i + k][j]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i - k][j]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i + k][j + k]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i - k][j - k]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i - k][j + k]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    cnt = 0
    try:
        for k in range(2, 8, 2):
            if arr[i][j] == arr[i + k][j - k]:
                cnt += 1
        if cnt == 3:
            return 1
    except IndexError:
        pass
    return 0

def isEnd(arr, n, m):
    for i in range(1, n, 2):
        for j in range(1, m, 2):
            if arr[i][j] == ' ●':
                if isLine(arr, i, j):
                    return 1
            elif arr[i][j] == ' ○':
                if isLine(arr, i, j):
                    return 2
    return 0
    
def clear(arr, n, m):
    for i in range(n):
        if i % 2:
            arr[i][0] = '│'
            arr[i][m - 1] = '│'
        else:
            arr[i][0] = '├'
            arr[i][m - 1] = '┤'
        for j in range(m):
            if i % 2 and not j % 2:
                arr[i][j] = '│'
        for j in range(m):
            if not j % 2:
                arr[0][j] = '┬'
                arr[n - 1][j] = '┴'
            else:
                arr[0][j] = '───'
                arr[n - 1][j] = '───'
                if not i % 2:
                    arr[i][j] = '───'
                else:
                    if i < n - 1 and j < m - 1:
                        arr[i + 1][j + 1] = '┼'
    arr[0][0], arr[0][m - 1], arr[n - 1][0], arr[n - 1][m - 1] = '┌', '┐', '└', '┘'

def connect4():
    n = 2 * 6 + 1
    m = 2 * 7 + 1
    height = [[n - 2] * 2 for i in range(7)]
    arr = [["   "] * m for i in range(n)]
    clear(arr, n, m)
    while True:
        for i in range(0, 2):
            printLayout(arr, n, m)
            p = input(f"플레이어 {i + 1}, 놓을 곳의 위치를 적으세요: ")
            while True:
                try:
                    p = int(p)
                    while p < 1 or p > 7:
                        printLayout(arr, n, m)
                        p = input(f"놓을 곳의 위치를 1 ~ 7의 범위로 적으세요: ")
                        p = int(p)
                    while height[p - 1][i] == -1:
                        printLayout(arr, n, m)
                        p = input(f"꽉 찼습니다. 다른 숫자를 입력하세요: ")
                        p = int(p)
                    break
                except BaseException:
                    printLayout(arr, n, m)
                    p = input(f"놓을 곳의 위치를 숫자로 적으세요: ")
            y = 2 * (p - 1) + 1
            x = height[p - 1][i]
            if i == 0:
                arr[x][y] = ' ●' 
            else:
                arr[x][y] = ' ○'
            height[p - 1][0] -= 2
            height[p - 1][1] -= 2
            if isFull(arr, n, m) and not isEnd(arr, n, m):
                printLayout(arr, n, m)
                print("무승부")
                return 0
            elif isEnd(arr, n, m):
                printLayout(arr, n, m)
                print(f"게임 끝 승자: {isEnd(arr, n, m)}")
                return 0
            else:
                pass
        
@choice
def connect4_main():
    printLogo()
    player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    while player_mode != "1" and player_mode != "2":
        printLogo()
        print("잘못 입력하셨습니다. 다시 입력해 주세요.")
        player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    else:
        if player_mode == "1":
            clear_screen()
            connect4()
            return 0
        elif player_mode == "2":
            return 1

if __name__ == "__main__":
    connect4_main()
    input()



br31_main.py

import os

#아래의 출처에서 코드를 수정함
#출처: https://stackoverflow.com/questions/22885780/python-clear-the-screen
from platform   import system as system_name 
def clear_screen(): 
    command = 'cls' if system_name().lower().startswith('win') else 'clear'
    os.system(command)

def choice(func):
    def wrapper():
        if func():
            clear_screen()
            return 0
        ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        while ch != '1' and ch != '2':
            print('다시 입력하세요.')
            ch = input('[1. 다시 시작] [2. 메인 메뉴로]\n선택지를 입력: ')
        else:
            if ch == '1':
                clear_screen()
                @choice
                def func1():
                    func()
                return func1()
            else:
                clear_screen()
    return wrapper

def printLogo():
    clear_screen()
    print(" _            _____ _")
    print("| |__  _ __  |___ // |")
    print("| '_ ⧵| '__|   |_ ⧵| |")
    print("| |_) | |     ___) | |")
    print("|_.__/|_|    |____/|_|")
    print("")
    
@choice
def br31_main():
    now = 0
    printLogo()
    player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    while player_mode != "1" and player_mode != "2":
        clear_screen()
        printLogo()
        print("잘못 입력하셨습니다. 다시 입력해 주세요.")
        player_mode = input("[1. 게임 플레이] [2. 메인 화면]\n선택지를 입력: ")
    else:
        if player_mode == "1":
            clear_screen()
            br31()
            return 0
        elif player_mode == "2":
            return 1

now = 0

def br31():
    global now
    printLogo()
    ch = input("플레이어 수를 입력하세요: ")
    while True:
        try:
            ch = int(ch)
            break
        except BaseException:
            printLogo()
            ch = input("플레이어 수를 숫자로 입력하세요: ")
    while True:
        for i in range(ch):
            printLogo()
            print(f"현재 숫자: {now}")
            num = input(f"플레이어{i + 1}, 1 ~ 3을 입력하세요: ")
            while True:
                try:
                    num = int(num)
                    if not(1 <= num and num <= 3):
                        num = input("1 ~ 3을 입력하세요: ")
                    else:
                        break
                except BaseException:
                    ch = input("숫자로 입력하세요: ")
            now += num
            if now >= 31:
                print(f"진 플레이어: {i + 1}")
                break
        if now >= 31:
            break
    now = 0
        
if __name__ == "__main__":
    br31_main()
